### ISSUE TITLE
An option/filter/callback_plugin to display variable inheritance path.

### ISSUE THREAD

**Issue by flare-ws** [ansible/ansible#25904]
##### ISSUE TYPE

 - Feature Idea

##### COMPONENT NAME
core, mb jinja filter, mb callback plugin

##### ANSIBLE VERSION
```
ansible 2.4.0 (devel 3aa41eda0b)
```

##### CONFIGURATION
N/A
##### OS / ENVIRONMENT
N/A

##### SUMMARY
Synopsis:
There are a LOT of places where a variable can be redefined during play. Here they are:
```
role defaults [1]
inventory file or script group vars [2]
inventory group_vars/all
playbook group_vars/all
inventory group_vars/*
playbook group_vars/*
inventory file or script host vars [2]
inventory host_vars/*
playbook host_vars/*
host facts
play vars
play vars_prompt
play vars_files
role vars (defined in role/vars/main.yml)
block vars (only for tasks in block)
task vars (only for the task)
role (and include_role) params
include params
include_vars
set_facts / registered vars
extra vars (always win precedence)
```
Although best practices tells us to minimize overrides count, in complex environments, we may need to redefine variable multiple times. Actually, if we have list of variables and some of them were dynamically overriden twice or more times (stage -> test configuration -> host or something), its very hard to tell where did a value came from.

A person who does not know ansible well will not be able to easily say where did a value comes from in multiple inheritance environment.

It will be a great option to see an **inheritance path** for a value. 

##### EXPECTED RESULTS
Inheritance display should be invoked either as jinja filter:
```
- debug:
    msg: "{{ some_variable | print_inheritance_path }}"
```
```
    var: "some_variable"
    path: "some_role/defaults.yml:1, group_vars/dev.yml:22, group_vars/test_scheme.yml:4, some_role/tasks/main.yml/lookup:222" 
```

Or in callback plugin (preferred):

```
TASK: [template]
...
...{ "action": "template", "vars": {
    "some_variable": [
        {
        "some_role/defaults.yml": "role default",
        "line": 1
         },
        {
        "group_vars/dev.yml": "value for the whole dev env",
        "line": 22
         },
        {
        "group_vars/test_scheme.yml": "value for test scheme",
        "line": 4
         },
        {
        "some_role/tasks/main.yml/lookup": "value acquired via dynamic lookup in playbook",
        "line": 222
         }
    ]}

```

##### ACTUAL RESULTS
N/A

**Comment by jctanner** [comment]
```
<@jtanner> James Tanner flare: i could swear alikins has been working on something related to that recently
10:06 AM <@alikins> jtanner, flare: I have a branch that has a bit of that
10:07 AM <flare> 109.235.164.18 - http://webchat.freenode.net Is there a way I can help you with it? :D
10:10 AM <@alikins> flare, https://github.com/ansible/ansible/compare/devel...alikins:var_man_changed_show_change  is one of them (I'll rebase to devel shortly)
```

**Comment by alikins** [comment]
The branch in question: https://github.com/ansible/ansible/compare/devel...alikins:var_man_changed_show_change

example output at https://gist.github.com/alikins/405352d8521ce8792fc2f72fde26f9ef

**Comment by flare-ws** [comment]
Suggested output for --diff mode (without -vvvvv):
https://gist.github.com/flare-ws/29da6fbc335aa93b05cbd220d8ec0278

**Comment by sakaal** [comment]
I know many people who would need this feature, but would not know how to ask for it, i.e. they would not easily find that this feature idea has already been proposed.

I certainly second that we need this, and I would add easy instructions on how to use it in the Ansible documentation page that talks about variables in general -- specifically in the part where the order of inheritance is discussed. That way everyone would find it easily.

**Comment by relrod** [comment]
Closed by #66102.
